crates/core/tedge_mapper/src/c8y/converter.rs

```diff
@@ -1,43 +1,78 @@
-use crate::c8y_fragments::{C8yAgentFragment, C8yDeviceDataFragment};
-use crate::error::*;
-use crate::size_threshold::SizeThreshold;
-use crate::{converter::*, operations::Operations};
-use c8y_smartrest::smartrest_serializer::{SmartRestSerializer, SmartRestSetSupportedOperations};
-use c8y_smartrest::{alarm, event};
+use crate::core::{converter::*, error::*, operations::Operations, size_threshold::SizeThreshold};
+use agent_interface::{
+    topic::{RequestTopic, ResponseTopic},
+    Auth, DownloadInfo, Jsonify, OperationStatus, RestartOperationRequest,
+    RestartOperationResponse, SoftwareListRequest, SoftwareListResponse, SoftwareUpdateResponse,
+};
+use async_trait::async_trait;
+use c8y_smartrest::{
+    alarm,
+    error::SmartRestDeserializerError,
+    event::serialize_event,
+    smartrest_deserializer::{SmartRestRestartRequest, SmartRestUpdateSoftware},
+    smartrest_serializer::{
+        CumulocitySupportedOperations, SmartRestGetPendingOperations, SmartRestSerializer,
+        SmartRestSetOperationToExecuting, SmartRestSetOperationToFailed,
+        SmartRestSetOperationToSuccessful, SmartRestSetSupportedLogType,
+        SmartRestSetSupportedOperations,
+    },
+};
 use c8y_translator::json;
-use mqtt_channel::{Message, Topic};
-use std::collections::hash_map::Entry;
-use std::collections::{HashMap, HashSet};
-use std::fs::File;
-use std::io::Read;
-use std::path::Path;
-use thin_edge_json::alarm::ThinEdgeAlarm;
-use thin_edge_json::event::ThinEdgeEvent;
-use tracing::info;
+use mqtt_channel::{Message, Topic, TopicFilter};
+use std::{
+    collections::{hash_map::Entry, HashMap, HashSet},
+    fs::File,
+    io::Read,
+    path::Path,
+    process::Stdio,
+};
+use thin_edge_json::{alarm::ThinEdgeAlarm, event::ThinEdgeEvent};
+use tracing::{debug, info, log::error};
+
+use super::{
+    error::SMCumulocityMapperError,
+    fragments::{C8yAgentFragment, C8yDeviceDataFragment},
+    http_proxy::C8YHttpProxy,
+    json_c8y::C8yUpdateSoftwareListResponse,
+    mapper::CumulocityMapper,
+    topic::{C8yTopic, MapperSubscribeTopic},
+};
 
 const C8Y_CLOUD: &str = "c8y";
 const INVENTORY_FRAGMENTS_FILE_LOCATION: &str = "/etc/tedge/device/inventory.json";
-const INVENTORY_MANAGED_OBJECTS_TOPIC: &str = "c8y/inventory/managedObjects/update";
 const SUPPORTED_OPERATIONS_DIRECTORY: &str = "/etc/tedge/operations";
+const INVENTORY_MANAGED_OBJECTS_TOPIC: &str = "c8y/inventory/managedObjects/update";
 const SMARTREST_PUBLISH_TOPIC: &str = "c8y/s/us";
 const TEDGE_ALARMS_TOPIC: &str = "tedge/alarms/";
 const INTERNAL_ALARMS_TOPIC: &str = "c8y-internal/alarms/";
-const TEDGE_MEASUREMENTS_TOPIC: &str = "tedge/measurements";
-const TEDGE_EVENTS_TOPIC: &str = "tedge/events/";
 
-pub struct CumulocityConverter {
+#[derive(Debug)]
+pub struct CumulocityConverter<Proxy>
```

Comment:
Curious to know why `CumulocityConverter` was made generic with `Proxy` instead of a the `http_proxy` being just an internal field of this struct(a `Box<dyn C8YHttpProxy>`) ?

Reply:
Compile time dispatch.